include/net/dns_sd.h

```diff
@@ -0,0 +1,130 @@
+extern "C" {
+#endif
+
+/**
+ * Register a service for DNS Service Discovery
+ *
+ * Example (no TXT):
+ * @code{c}
+ * #include <net/dns_sd.h>
+ * static const foo_txt[] = DNS_SD_EMPTY_TXT;
+ * DNS_SD_REGISTER_SERVICE(foo, "My Foo", "_foo", "_tcp", foo_txt, 4242);
+ * @endcode{cpp}
+ *
+ * TXT records begin with a single length byte (hex-encoded)
+ * and contain key=value pairs.
+ *
+ * For additional rules on TXT encoding, see RFC 6763, Section 6.4.
+ *
+ * Example (with TXT):
+ * @code{c}
+ * #include <net/dns_sd.h>
+ * static const bar_txt[] = {
+ *   "\x06" "path=/"
+ *   "\x0f" "this=is the way"
+ *   "\x0e" "foo or=foo not"
+ *   "\x17" "this=has\0embedded\0nulls"
+ *   "\x04" "true"
+ * };
+ * DNS_SD_REGISTER_SERVICE(bar, "My Bar", "_bar", "_tcp", bar_txt, 0x4242);
+ * @endcode{cpp}
+ */
+#define DNS_SD_REGISTER_SERVICE(id, instance, service, proto, text, port) \
+    static const struct _dns_sd_rec id \
+        __attribute__((used)) __attribute__((section("dns_sd_rec"))) = { \
+        instance, \
+        service, \
+        proto, \
+		"local", \
+		(const char *)text, \
+		sizeof(text) - 1, \
+        port \
+    }
+
+#define DNS_SD_REGISTER_TCP_SERVICE(id, instance, service, text, port) \
+    DNS_SD_REGISTER_SERVICE(id, instance, service, "_tcp", text, port)
+
+#define DNS_SD_REGISTER_UDP_SERVICE(id, instance, service, text, port) \
+    DNS_SD_REGISTER_SERVICE(id, instance, service, "_udp", text, port)
+
+/**
+ * Calculate the size of the DNS-SD service 
+ * 
+ * This macro calculates the size of the DNS-SD service for a DNS
+ * Resource Record (RR).
+ * 
+ * For example, if there is a service called 'My Foo'._http._tcp.local.,
+ * then the returned size is 18. That is broken down as shown below.
+ * 
+ * - 1 byte for the size of "_http"
+ * - 5 bytes for the value of "_http"
+ * - 1 byte for the size of "_tcp"
+ * - 4 bytes for the value of "_tcp"
+ * - 1 byte for the size of "local"
+ * - 5 bytes for the value of "local"
+ * - 1 byte for the trailing NUL terminator '\0' 
+ */
+#define DNS_SD_SERVICE_PROTO_SIZE(ref) \
+    (1 + strlen(ref->service) + 1 + strlen(ref->proto) + 1 + strlen(inst->domain) + 1)
+
+/**
+ * Calculate the size of a DNS-SD name
+ * 
+ * RFC 1035, Section 2.3.5 requires that DNS names (including delimiters)
+ * to be no more than 255 characters.
+ * 
+ * This function calculates the length of the DNS-SD name.
+ * 
+ * For example, if there is a service called 'My Foo'._http._tcp.local.,
+ * then the returned size is 24. That is broken down as shown below.
+ * 
+ * - 6 bytes for the value of "My Foo"
+ * - 1 byte for a delimiter
+ * - 5 bytes for the value of "_http"
+ * - 1 byte for a delimiter
+ * - 4 bytes for the value of "_tcp"
+ * - 1 byte for a delimiter
+ * - 5 bytes for the value of "local"
+ * - 1 byte for the trailing NUL terminator '\0' 
+ */
+#define DNS_SD_NAME_SIZE(ref) \
+    (strlen(ref->instance) + 1 + strlen(ref->service) + 1 + strlen(ref->proto) + 1 + strlen("local") + 1)
```

Comment:
This should actually be calculated at runtime, I think. At least I'm using it that way. Probably better to make it a function to be honest, rather than a macro, just so that isn't a point of confusion.